File: cjaasNodeJSClient/tools/generatesastoken.js
"use strict";
const { ArgumentParser } = require("../apis/node_modules/argparse");
const { version } = require("../apis/package.json");
const crypto = require("crypto");

const services = [
  ["profile", "r", "profileReadToken"],
  ["profile", "w", "profileWriteToken"],
  ["stream", "r", "streamToken"],
  ["tape", "r", "tapeToken"],
  ["ds", "w", "datasinkWriteToken"],
  ["idmt", "r", "identityReadToken"],
  ["idmt", "w", "identityWriteToken"],
  ["action", "w", "actionWriteToken"],
  ["action", "r", "actionReadToken"],
  ["account", "r", "accountReadToken"],
];

const servicesForGadget = ["profile", "stream", "tape", "idmt"];

function generateExpiration(args) {
  var validitySum = 0;
  validitySum += parseInt(args["vd"]) ? args["vd"] * 60 * 60 * 24 : 0;
  validitySum += parseInt(args["vh"]) ? args["vh"] * 60 * 60 : 0;

  if (validitySum === 0) {
    //default 1 week
    validitySum = 60 * 60 * 24 * 7;
  }

  var now = new Date();
  return Math.round(now.getTime()) + validitySum * 1000;
}

function verifyArguments() {
  const parser = new ArgumentParser({
    description: "SAS Token Generator!",
  });

  parser.add_argument("-v", "--version", { action: "version", version });
  parser.add_argument("-secret", { help: "secret: your secret key" });
  parser.add_argument("-o", { help: "organization: organization" });
  parser.add_argument("-n", { help: "namespace: customer namespace" });
  parser.add_argument("-s", {
    help: "service: particular service to be queried",
  });
  parser.add_argument("-p", {
    help: "permissions: one of the following (r|w|rw)",
  });
  parser.add_argument("-kn", { help: "keyname: key name" });
  parser.add_argument("-vd", {
    help: "validitydays: expire in this number of days",
  });
  parser.add_argument("-vh", {
    help: "validityhours:expire in this number of hours",
  });
  parser.add_argument("-a", {
    help: "generate all tokens for JDS",
    action: "store_true",
  });
  parser.add_argument("-g", {
    help: "generate query string for Customer Journey Finesse Gadget",
    action: "store_true",
  });

  var args = parser.parse_args();

  if (!args["secret"]) {
    throw new Error("-secret argument missing!");
  }
  if (!args["o"]) {
    throw new Error("-o argument missing!");
  }
  if (!args["n"]) {
    throw new Error("-n argument missing!");
  }
  if (!args["s"] && !args["a"] && !args["g"]) {
    throw new Error("-s argument missing!");
  }
  if (!args["p"] && !args["a"] && !args["g"]) {
    throw new Error("-p argument missing!");
  }
  if (!args["kn"]) {
    throw new Error("-kn argument missing!");
  }

  return args;
}
function generateSignature(secret, message) {
  return encodeURIComponent(
    crypto.createHmac("sha256", secret).update(message).digest("base64"),
  );
}

function generateQueryString(queryStringArray) {
  let queryString = "?";

  queryStringArray.forEach((listItem, i) => {
    queryString += `${listItem[0]}=${encodeURIComponent(listItem[1])}${
      i === queryStringArray.length - 1 ? "" : "&"
    }`;
  });

  return queryString;
}

function generateSasToken(
  secret,
  organization,
  namespace,
  service,
  permission,
  keyName,
  expiration,
) {
  var sasTokenPrefix = `so=${organization}&sn=${namespace}&ss=${service}&sp=${permission}&se=${expiration}&sk=${keyName}`;

  const signature = generateSignature(secret, sasTokenPrefix);

  return `${sasTokenPrefix}&sig=${signature}`;
}

function generateSasTokens(args) {
  var secret = args["secret"];
  var organization = args["o"];
  var namespace = args["n"];
  var service = args["s"];
  var permission = args["p"];
  var keyName = args["kn"];
  var expiration = new Date(generateExpiration(args)).toISOString();

  if (args["a"] || args["g"]) {
    const queryStringArray = [];
    services.forEach((serviceCombination) => {
      const _service = serviceCombination[0];

      const _permission = serviceCombination[1];

      const token = generateSasToken(
        secret,
        organization,
        namespace,
        _service,
        _permission,
        keyName,
        expiration,
      );

      if (args["a"]) {
        console.log(
          `\n======================SAS Token for ${_service} with ${_permission} permission=======================================`,
        );
        console.log(`SharedAccessSignature ${token}`);
      }

      if (
        args["g"] &&
        servicesForGadget.indexOf(serviceCombination[0]) !== -1
      ) {
        queryStringArray.push([serviceCombination[2], token]);
      }
    });

    if (args["g"]) {
      const queryString = generateQueryString(queryStringArray);
      console.log(
        `\n======================Query String for Customer Journey Finesse Gadget=======================================`,
      );
      console.log(queryString);
    }
  } else {
    const token = generateSasToken(
      secret,
      organization,
      namespace,
      service,
      permission,
      keyName,
      expiration,
    );
    return `SharedAccessSignature ${token}`;
  }
}

/* Begin */
try {
  var verifiedArguments = verifyArguments();
  generateSasTokens(verifiedArguments);
} catch (err) {
  console.log("\n-- ERROR --");
  console.log(err);
}

File: cjaasPythonClient/tools/generate_sas_token.py
import base64
import hashlib
import hmac
import urllib.parse
from argparse import ArgumentParser
from datetime import datetime, timedelta

services = (
        ["profile", "r", "profileReadToken"],
        ["profile", "w", "profileWriteToken"],
        ["stream", "r", "streamToken"],
        ["tape", "r", "tapeToken"],
        ["ds", "w", "datasinkWriteToken"],
        ["idmt", "r", "identityReadToken"],
        ["idmt", "w", "identityWriteToken"],
        ["action", "w", "actionWriteToken"],
        ["action", "r", "actionReadToken"],
        ["account", "r", "accountReadToken"],
)

services_for_gadget = ("profile", "stream", "tape", "idmt")


def parse_command_line_args():
    arg_parser = ArgumentParser()
    arg_parser.add_argument("--secret", required=True, default=None,
                            help="secret: your secret key")
    arg_parser.add_argument("--o", required=False, default="123",
                            help="organization: organization")
    arg_parser.add_argument("--n", required=False, default="456",
                            help="namespace: customer namespace")
    arg_parser.add_argument("--s", required=False, default=None,
                            help="service: particular service to be queried")
    arg_parser.add_argument("--p", required=False, default=None,
                            help="permissions: one of the following (r|w|rw)")
    arg_parser.add_argument("--kn", required=False, default="orc",
                            help="keyname: key name")
    arg_parser.add_argument("--vd", required=False, default=365,
                            help="validitydays: expire in this number of days", type=int)
    arg_parser.add_argument("--vh", required=False, default=1,
                            help="validityhours:expire in this number of hours", type=int)
    arg_parser.add_argument("--g", required=False, default=False,
                            help="for-gadget will generate query string for Customer Journey Gadget", type=bool)
    return arg_parser.parse_args()

def generate_signature(token, param_secret):
    message = bytes(token, 'utf-8')
    secret = bytes(param_secret, 'utf-8')
    signature = base64.b64encode(
        hmac.new(secret, message, digestmod=hashlib.sha256).digest())
    result = token + \
        "&sig=" + urllib.parse.quote_plus(signature.decode("utf-8"))
    
    return result


class generate_saas_token:

    @classmethod
    def generate_token(cls, param):
        # print('Args passed in to the code: {}'.format(param))
        if param.secret == None:
            print('\n Missing secret to generate SAS Tokens')
            return
        if param.vd == 0 and param.vh == 0:
            raise ValueError("Both validityDays and validityHours can't be 0")
        valid_until = (datetime.utcnow() + timedelta(days=param.vd) + timedelta(
            hours=param.vh)).isoformat(timespec='milliseconds')

        if param.s == None or param.p == None:
            query_string = []
            for x in services:
                token= "so=" + param.o + "&sn=" + param.n + "&ss=" + x[0] + \
                    "&sp=" + x[1] + "&se=" + \
                    valid_until + "Z&sk=" + param.kn

                result = generate_signature(token, param.secret)
                print('\n-----------------------------------SAS Token for ' + x[0] + ' with '+ x[1] + ' permission------------------------')
                print('\nSharedAccessSignature ' + result)
                if x[0] in services_for_gadget:
                    query_string.append([x[2], result])
                
            if param.g == True:
                query = '?'
                index = 0
                for tup in query_string:
                    query += tup[0]  + '=' + urllib.parse.quote_plus(tup[1])
                    index += 1
                    if index != len(query_string):
                        query += '&' 
                print('\n-----------------------------------Query String for Customer Journey Gadget------------------------')
                print('\n' + query)
        else:
            token = "so=" + param.o + "&sn=" + param.n + "&ss=" + param.s + \
                "&sp=" + param.p + "&se=" + \
                valid_until + "Z&sk=" + param.kn
            result = generate_signature(token, param.secret)
            print('\n-----------------------------------SAS Token for ' + param.s + ' with '+ param.p + ' permission------------------------')
            print('\nSharedAccessSignature ' + result)

        return result

if __name__ == "__main__":
    generate_saas_token.generate_token((parse_command_line_args()))
